include/s3select_json_parser.h

```diff
@@ -0,0 +1,446 @@
+    ChunksStreamer stream_buffer;
+    bool init_buffer_stream;
+    rapidjson::Reader reader;
+    std::vector<en_json_elm_state_t> json_element_state;
+    std::string m_result;//debug purpose
+    std::vector<std::string> key_path;
+    std::string m_filter_result{};
+
+    JsonParserHandler() : init_buffer_stream(false)
+    {}
+
+    std::string get_key_path()
+    {//for debug
+	std::string res;
+	for(const auto & i: key_path)
+	{
+	  res.append(i);
+	  res.append(std::string("/"));
+	}
+	return res;
+    }
+
+    void clear_filtermap() {
+      filtermap.clear();
+    }
+
+    std::vector < std::pair < std::string, Valuesax>> get_filtervalue() {
+      return filtermap;
+    }
+
+    void dec_key_path()
+    {
+      if(json_element_state.back() != ARRAY_STATE)
+	{
+	  if(key_path.size() != 0)
+	    key_path.pop_back();
+	}
+    }
+
+    void push_new_key_value(Valuesax& v)
+    {
+
+      if(key_path.size() <= filter_pattern.size() && key_path != filter_pattern) {
+        flag = false;
+      }
+
+      if(key_value_criteria) {
+        filtermap.push_back(std::make_pair(get_key_path(), v));
+      }
+      
+      if (flag && filter_pattern_criteria) {
+        filtermap.push_back(std::make_pair(get_key_path(), v));
+      }
+      if (key_path == filter_exact && filter_exact_criteria) {
+        filtermap.push_back(std::make_pair(get_key_path(), v));
+      }
+
+      if (key_path == filter_prefix_suffix && prefix_suffix_exact_criteria) {
+        filtermap.push_back(std::make_pair(get_key_path(), v));
+      }
+      dec_key_path();
+    }
+
+    bool Null() {
+      // at this point should verify against from-clause/where-clause/project. if match then push to scratch-area
+      push_new_key_value(value.Parse(nullptr));
+      return true; }
+
+    bool Bool(bool b) {
+      push_new_key_value(value.Parse(b));
+      return true; }
+
+    bool Int(int i) { 
+      push_new_key_value(value.Parse(i));
+      return true; }
+
+    bool Uint(unsigned u) {
+      push_new_key_value(value.Parse(u));
+      return true; }
+
+    bool Int64(int64_t i) { 
+      push_new_key_value(value.Parse(i));
+      return true; }
+
+    bool Uint64(uint64_t u) { 
+      push_new_key_value(value.Parse(u));
+      return true; }
+
+    bool Double(double d) { 
+      push_new_key_value(value.Parse(d));
+      return true; }
+
+    bool String(const char* str, rapidjson::SizeType length, bool copy) {
+      //TODO use copy
+      push_new_key_value(value.Parse(str));
+      return true;
+    }
+
+    bool Key(const char* str, rapidjson::SizeType length, bool copy) {
+      key_path.push_back(std::string(str));
+      
+      if (key_path == filter_pattern) {
+        flag = true;
+      }
+      return true;
+    }
+
+    bool StartObject() {
+      json_element_state.push_back(OBJECT_STATE);
+      return true; 
+    }
+  
+    bool EndObject(rapidjson::SizeType memberCount) {
+      json_element_state.pop_back();
+      dec_key_path();
```

Comment:
please note that `dec_key_path` alters the key-path, and it is called by several callbacks.